backend/unparsed_value.go

```diff
@@ -42,6 +42,47 @@ type UnparsedVariableValue interface {
 // that were successfully processed, allowing for careful analysis of the
 // partial result.
 func ParseVariableValues(vv map[string]UnparsedVariableValue, decls map[string]*configs.Variable) (terraform.InputValues, tfdiags.Diagnostics) {
+	ret, diags := ParseVariableValuesWithoutDefaults(vv, decls)
+
+	// Populate any variables we haven't gathered as their defaults and fail if any of
+	// the missing ones are required.
+	for name, vc := range decls {
+		if _, defined := ret[name]; defined {
+			continue
+		}
+
+		if vc.Required() {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "No value for required variable",
+				Detail:   fmt.Sprintf("The root module input variable %q is not set, and has no default value. Use a -var or -var-file command line argument to provide a value for this variable.", name),
+				Subject:  vc.DeclRange.Ptr(),
+			})
+
+			// We'll include a placeholder value anyway, just so that our
+			// result is complete for any calling code that wants to cautiously
+			// analyze it for diagnostic purposes. Since our diagnostics now
+			// includes an error, normal processing will ignore this result.
+			ret[name] = &terraform.InputValue{
+				Value:       cty.DynamicVal,
+				SourceType:  terraform.ValueFromConfig,
+				SourceRange: tfdiags.SourceRangeFromHCL(vc.DeclRange),
+			}
+		} else {
+			ret[name] = &terraform.InputValue{
+				Value:       vc.Default,
+				SourceType:  terraform.ValueFromConfig,
+				SourceRange: tfdiags.SourceRangeFromHCL(vc.DeclRange),
+			}
+		}
+	}
+	return ret, diags
+}
+
+// ParseVariableValuesWithoutDefaults performs the same operation as
+// ParseVariableValues, but does not set any default values or return any
+// errors for variables which are unset.
+func ParseVariableValuesWithoutDefaults(vv map[string]UnparsedVariableValue, decls map[string]*configs.Variable) (terraform.InputValues, tfdiags.Diagnostics) {
```

Comment:
I think it makes more sense to put this function in  `backend/remote` since it's very much specific to that backend, and leave this file unchanged. What do you think?

Reply:
This is actually used up on line 45 in this file too (for local runs), so I think it makes sense to keep it.

It is basically logic that was already in this file, but Alisdair extracted it into its own function so it could be consumed by remote.